client/src/entrypoints/admin/page-editor.js

```diff
@@ -402,6 +384,104 @@ window.updateFooterSaveWarning = (formDirty, commentsDirty) => {
+    'button[data-preview-size]',
+  );
+
+  const togglePreviewSize = (event) => {
+    const currentButton = event.currentTarget;
+
+    previewPanel.dataset.activePreviewSize = currentButton.dataset.previewSize;
+    previewButtons.forEach((b) => b.setAttribute('aria-current', 'false'));
+    currentButton.setAttribute('aria-current', 'true');
+  };
+
+  previewButtons.forEach((button) => {
+    button.addEventListener('click', togglePreviewSize);
+  });
+
+  const iframe = previewPanel.querySelector('iframe');
+  const refreshButton = previewPanel.querySelector('.refresh-button');
+  const openPreviewButton = previewPanel.querySelector('.open-preview-button');
+  const form = document.getElementById('page-edit-form');
+  const previewUrl = previewPanel.dataset.action;
+  const submitAction = document.querySelector('.action-save');
+
+  const setPreviewData = () =>
+    fetch(previewUrl, {
+      method: 'POST',
+      body: new FormData(form),
+    }).then((response) =>
+      response.json().then((data) => {
+        if (data.is_valid) {
+          previewPanel.classList.remove('has-errors');
+        } else {
+          previewPanel.classList.add('has-errors');
+        }
+        iframe.contentWindow.location.reload();
+        return data.is_valid;
+      }),
+    );
+
+  const handlePreview = () =>
+    setPreviewData()
+      .then((valid) => {
+        // Submit the form to show the validation errors.
+        if (!valid) submitAction.click();
+        return valid;
+      })
+      .catch(() => {
+        // eslint-disable-next-line no-alert
+        alert('Error while sending preview data.');
+      });
+
+  const handlePreviewInNewTab = () => {
+    const previewWindow = window.open('', previewUrl);
+    previewWindow.focus();
+
+    handlePreview().then((success) => {
+      if (success) {
+        previewWindow.document.location = previewUrl;
+      } else {
+        window.focus();
+        previewWindow.close();
+      }
+    });
+  };
+
+  refreshButton.addEventListener('click', handlePreview);
+  openPreviewButton.addEventListener('click', handlePreviewInNewTab);
+
+  // Make sure current preview data in session exists and is up-to-date.
+  setPreviewData();
+
+  if (previewPanel.dataset.autoUpdate === 'true') {
+    // Form data is changed when field values are changed (change event),
+    // when HTML elements are added, modified, moved, and deleted
+    // (childList changes observed by MutationObserver), and we need to delay
+    // setPreviewData when typing to avoid useless extra AJAX requests
+    // (so we postpone setPreviewData when keyup occurs).
+
+    let autoUpdatePreviewDataTimeout = -1;
+    const autoUpdatePreview = () => {
+      clearTimeout(autoUpdatePreviewDataTimeout);
+      autoUpdatePreviewDataTimeout = setTimeout(setPreviewData, 1000);
+    };
+
+    const observer = new MutationObserver(autoUpdatePreview);
+    observer.observe(form, { subtree: true, childList: true });
+
```

Comment:
Is the mutation observer necessary? Might be a bit of overkill for tracking changes. 

I would think that the keyboard and change event listeners might be enough.

Reply:
Yep, I agree it's a bit overkill. I just tried to replicate the existing behaviour here:

https://github.com/wagtail/wagtail/blob/c7a0d6ef9509317ca1f40a5095584f9f6cfe49d1/client/src/entrypoints/admin/page-editor.js#L325-L331

I'd be happy to drop it!

Reply:
ahh I did not see that.

Yeah see how we go without out I think, there are some 'change' events that will not really get triggered for things like inline panel additions.

However, technically we should probably trigger those events on the relevant fields.